src/pkcs11module.cpp

```diff
@@ -250,27 +241,34 @@ CK_RV PKCS11Module::sign(const std::vector<unsigned char> &cert, const std::vect
         return CKR_OBJECT_HANDLE_INVALID;
     }
 
-    CK_MECHANISM mechanism = {CKM_RSA_PKCS, 0, 0};
+    CK_KEY_TYPE keyType = CKK_RSA;
+    CK_ATTRIBUTE attribute = { CKA_KEY_TYPE, &keyType, sizeof(keyType) };
+    check_C(GetAttributeValue, session, key[0], &attribute, 1);
+
+    CK_MECHANISM mechanism = {keyType == CKK_ECDSA ? CKM_ECDSA : CKM_RSA_PKCS, nullptr, 0};
     check_C(SignInit, session, &mechanism, key[0]);
     std::vector<unsigned char> hashWithPadding;
-    // FIXME: explicit hash type argument
-    switch (hash.size()) {
-    case BINARY_SHA224_LENGTH:
-        hashWithPadding = {0x30, 0x2d, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x04, 0x05, 0x00, 0x04, 0x1c};
-        break;
-    case BINARY_SHA256_LENGTH:
-        hashWithPadding = {0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20};
-        break;
-    case BINARY_SHA384_LENGTH:
-        hashWithPadding = {0x30, 0x41, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x02, 0x05, 0x00, 0x04, 0x30};
-        break;
-    case BINARY_SHA512_LENGTH:
-        hashWithPadding = {0x30, 0x51, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x03, 0x05, 0x00, 0x04, 0x40};
-        break;
-    default:
-        _log("incorrect digest length, dropping padding");
+    if(keyType == CKK_RSA)
+    {
```

Comment:
Code style in this project seems to default to 
```
if(keyType == CKK_RSA) {
```

so having the braces after the if statement. I would prefer not to mix different coding styles.